### PR TITLE
Remove references to PlatformAbstractions.

### PR DESCRIPTION
**Pull request by eerhardt**
This will allow us to remove the PlatformAbstractions library from dotnet/runtime.

Contributes to https://github.com/dotnet/runtime/issues/3470

### CODE REVIEW

**Comment by eerhardt** [comment]
@dsplaisted @wli3 - thoughts on this?

**Comment by dsplaisted** [comment]
@eerhardt Is the RuntimeEnvironment class here a bunch of code copied from the PlatformAbstractions library?  Is there a way to avoid adding it here?

**Comment by eerhardt** [comment]
> Is the RuntimeEnvironment class here a bunch of code copied from the PlatformAbstractions library?

You can think of it as "moved" code from the PlatformAbstractions library. The main things that `dotnet/sdk` uses from `PlatformAbstractions` that we don't have APIs for in .NET 5 are:

1. `OperatingSystemPlatform`
2. `OperatingSystemVersion`
3. `OperatingSystem`

You can see the discussion about these APIs at https://github.com/dotnet/runtime/issues/26780.

> Is there a way to avoid adding it here?

The only way of avoiding it that I can think of is to change the callsites in `dotnet/sdk` to no longer use the above 3 APIs.

As discussed in https://github.com/dotnet/runtime/issues/26780#issuecomment-404565747, `dotnet/sdk` uses these APIs for logging/telemetry purposes, ex. `dotnet --info`. However, the `dotnet/runtime` team doesn't think these APIs are generally useful to developers and actually causes developers to use bad habits by checking if you are running on a specific version of an OS. Thus, the APIs were not approved.

Since `dotnet/runtime` doesn't want to expose these as APIs, the only options are:

1. Put the code here, where it is needed
2. Don't do it at all

So unless we want to change the telemetry information we are sending and change `dotnet --info`, moving the code here was my only option.

Another potential option would be to use [RuntimeInformation.OSDescription](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.runtimeinformation.osdescription?view=netframework-4.8) instead, but that would still change the format of the telemetry information and the `dotnet --info` information.

**Comment by dsplaisted** [comment]
@eerhardt How likely is it that this logic will need to be updated moving forward?  Who will be responsible for maintaining it?

Also it looks like it still has its own implementation of getting the RuntimeIdentifier, can we remove that code and use the one that's being added to the BCL?

File: src/Layout/toolset-tasks/toolset-tasks.csproj [hunk 1]
```diff
@@ -1,13 +1,17 @@
-<Project Sdk="Microsoft.NET.Sdk" ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+﻿<Project Sdk="Microsoft.NET.Sdk" ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <PropertyGroup>
     <TargetFrameworks>$(SdkTargetFramework);net472</TargetFrameworks>
     <TargetFrameworks Condition="'$(OS)' != 'Windows_NT'">$(SdkTargetFramework)</TargetFrameworks>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
   </PropertyGroup>
 
+  <ItemGroup>
+    <Compile Include="..\..\Cli\Microsoft.DotNet.Cli.Utils\RuntimeEnvironment.cs" />
```
Comment (wli3) [inline src/Layout/toolset-tasks/toolset-tasks.csproj:9]: This is not great. Directly source reference a file to production code. The build tasks have been moving around several times.

File: src/Cli/Microsoft.DotNet.Cli.Utils/RuntimeEnvironment.cs [hunk 2]
```diff
@@ -0,0 +1,372 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.IO;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.DotNet.Cli.Utils
```
Comment (wli3) [inline src/Cli/Microsoft.DotNet.Cli.Utils/RuntimeEnvironment.cs:8]: no test coverage

**Comment by wli3** [comment]
I am very concerned about this change considering the left over part - RuntimeEnvironment.cs is still so big.

Also there is mostly no test coverage for the entire file -- and it would be hard to test since it varies and requires testing on different machine config -- and that's why it is bad to be in SDK. SDK repo is not setup to test different combination of OSs. If I need to fix a bug in this code, it will be an edit and pray - and find a regression - and fix and pray again situation.

Some ideas:
1. Keep RuntimeEnvironment.cs in runtime repo and produce a library out of it, since it can be better tested there.
2. Add full test coverage to RuntimeEnvironment.cs somehow in SDK
3. RuntimeEnvironment.cs file will still be owned by runtime and log it by [this](https://help.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners#codeowners-file-location). Considering the repo merge, we no longer have a hard boundary  between team and repos, this is normal now.

**Comment by eerhardt** [comment]
> How likely is it that this logic will need to be updated moving forward? 

I can't really predict the future, but in the past 2.5 years, this has been the only substantive change: https://github.com/dotnet/core-setup/pull/4085

> Who will be responsible for maintaining it?

I'm fine with me or someone from the libraries/runtime team maintaining it. But I don't see it as different than the current [get libc version](https://github.com/dotnet/sdk/blob/b1223209644d900702287faea8e9b71f95ec49f8/src/Cli/dotnet/Telemetry/ExternalTelemetryProperties.cs#L91-L132) telemetry code we have today.

> Also it looks like it still has its own implementation of getting the RuntimeIdentifier, can we remove that code and use the one that's being added to the BCL?

That implementation is used by the [GetCurrentRuntimeInformation ](https://github.com/dotnet/sdk/blob/b1223209644d900702287faea8e9b71f95ec49f8/src/Layout/toolset-tasks/GetCurrentRuntimeInformation.cs#L24) task in `toolset-tasks.csproj`, which builds for both netcoreapp and net472. The new API in the BCL is only in netcoreapp.

> I am very concerned about this change considering the left over part - RuntimeEnvironment.cs is still so big.

If we can remove the telemetry and `dotnet --info` functionality that uses this code, we can remove this left over part. For example, do we really need both `OS Name` and `OS Platform` in `dotnet --info` and telemetry? The advantage of putting this code in `dotnet/sdk` where it is used is if/when we change the telemetry to no longer need this, we can just delete the code. Where if we have this functionality as an official API, we can never remove it.

It looks like some of the version code in RuntimeEnvironment.cs can be refactored to instead use `System.Environment.OSVersion`. That will make for less code here.

> Also there is mostly no test coverage for the entire file

I can add some tests for the functions. For the most part, there are only 3 main code paths: windows, osx, and linux.

> Keep RuntimeEnvironment.cs in runtime repo and produce a library out of it, since it can be better tested there.

This is not a viable solution. It is what we have today and what I'm trying to fix. The PlatformAbstractions library is in a weird "limbo" state, where it hasn't been officially approved, but it is in the same repo as the official BCL APIs.

File: src/Layout/toolset-tasks/toolset-tasks.csproj (hunk 1 above)
Comment (eerhardt) [reply src/Layout/toolset-tasks/toolset-tasks.csproj:9]: I don't know of a better way to do this. Both the production code, and this build task need this functionality. Do you have a recommendation?

> The build tasks have been moving around several times.

If it moves again, then this line will need to be updated. But the mover will get a compile error when they move it, so I don't see a problem.

**Comment by eerhardt** [comment]
@dsplaisted @wli3 - I believe I've addressed your feedback. Please let me know what you think.

File: src/Cli/Microsoft.DotNet.Cli.Utils/RuntimeEnvironment.cs (hunk 2 above)
Comment (eerhardt) [reply src/Cli/Microsoft.DotNet.Cli.Utils/RuntimeEnvironment.cs:8]: Added.

**Comment by wli3** [comment]
>  I don't see it as different than the current [get libc version](https://github.com/dotnet/sdk/blob/b1223209644d900702287faea8e9b71f95ec49f8/src/Cli/dotnet/Telemetry/ExternalTelemetryProperties.cs#L91-L132)

I actually made the same comment then that code is checked-in. And it is also maintained by runtime team

File: src/Cli/Microsoft.DotNet.Cli.Utils/RuntimeEnvironment.cs [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+        public static string OperatingSystemVersion { get; } = GetOSVersion();
+        public static string OperatingSystem { get; } = GetOSName();
+
+        // toolset-tasks.csproj needs to build for full .NET Framework, and needs to get the current
+        // RuntimeIdentifier of the machine. Because of this, it needs to implement
+        // it's own GetRuntimeIdentifier(). All other places should use RuntimeInformation.RuntimeIdentifier.
+#if TOOLSET_TASKS
+        private static readonly string OverrideEnvironmentVariableName = "DOTNET_RUNTIME_ID";
+
+        public static string GetRuntimeIdentifier()
```
Comment (dsplaisted) [inline src/Cli/Microsoft.DotNet.Cli.Utils/RuntimeEnvironment.cs:34]: It looks like we're not even using the [value that is returned](https://github.com/dotnet/sdk/blob/master/src/Layout/redist/targets/GetRuntimeInformation.targets#L6) in the `Rid` output parameter of the GetCurrentRuntimeInformation task.  So it looks like we actually can delete this implementation of `GetRuntimeIdentifier`.

**Review by dsplaisted — APPROVED** [review]
I think I am OK with this.  The use of this new code is restricted to telemetry and informational output.

As I commented, I think we can get rid of the duplicate GetRuntimeIdentifier implementation altogether.

File: src/Cli/Microsoft.DotNet.Cli.Utils/RuntimeEnvironment.cs (hunk 3 above)
Comment (eerhardt) [reply src/Cli/Microsoft.DotNet.Cli.Utils/RuntimeEnvironment.cs]: Nice catch! After seeing that isn't used, I was able to clean up a lot more code as well.

File: src/Layout/toolset-tasks/toolset-tasks.csproj (hunk 1 above)
Comment (eerhardt) [reply src/Layout/toolset-tasks/toolset-tasks.csproj]: This is now removed with the latest changes.

**Pull request merged**